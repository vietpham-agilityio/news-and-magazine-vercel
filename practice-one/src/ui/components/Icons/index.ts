export * from './User';
export * from './Plus';
export * from './Home';
export * from './File';
export * from './Close';
export * from './Circle';
export * from './Search';
export * from './SaveTag';
export * from './MailBox';
export * from './Twitter';
export * from './ChatBox';
export * from './Ellipse';
export * from './Schedule';
export * from './ArrowTop';
export * from './ThreeDots';
export * from './Instagram';
export * from './BurgerMenu';
export * from './ChevronsUp';
export * from './ArrowLeft';
export * from './ArrowRight';
export * from './PaperPlane';
export * from './ArrowBottom';
export * from './SaveTagSmall';