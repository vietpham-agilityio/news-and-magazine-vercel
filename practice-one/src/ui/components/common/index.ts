export * from './Tag';
export * from './Menu';
export * from './Avatar';
export * from './Button';
export * from './BoxIcon';
export * from './SubMenu';
export * from './Typography';
export * from './Information';
export * from './TitleSection';
