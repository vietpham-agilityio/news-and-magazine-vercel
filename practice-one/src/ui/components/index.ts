// common
export * from '@/ui/components/common';

// shared
export * from '@/ui/components/Video';
export * from '@/ui/components/Brand';
export * from '@/ui/components/NavItem';
export * from '@/ui/components/Comment';
export * from '@/ui/components/Carousel';
export * from '@/ui/components/SearchBar';
export * from '@/ui/components/PaperPost';
export * from '@/ui/components/PaperVideo';
export * from '@/ui/components/BurgerMenu';
export * from '@/ui/components/Card/CardSingle';
export * from '@/ui/components/Skeleton/CardPost';
export * from '@/ui/components/Skeleton/DetailPost';
export * from '@/ui/components/Skeleton/CardVertical';
export { default as CardPost } from '@/ui/components/Card/CardPost';