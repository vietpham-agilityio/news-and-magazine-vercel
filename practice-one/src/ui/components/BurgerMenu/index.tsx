'use client';

import { useState, useEffect } from 'react';

// constants
import { flexItemCenter } from '@/constants';

// mocks
import { listNavItems } from '@/mocks';

// components
import { BoxIcon, Brand, NavItem, Typography } from '@/ui/components';
import Link from 'next/link';

// icons
import {
  ArrowBottom,
  ArrowRight,
  Close,
  BurgerMenu as BurgerMenuIcon,
} from '@/ui/components/Icons';

// types
import { FontWeight, IMenu, Size } from '@/types';
import { createPortal } from 'react-dom';

interface IProps {
  listCategories: IMenu[];
}

export const BurgerMenu = ({ listCategories }: IProps) => {
  const [isOpenMenu, setIsOpenMenu] = useState<boolean>(false);
  const [isOpenSubMenu, setIsOpenSubMenu] = useState<boolean>(false);

  const classNavItem = `w-full h-12 pl-2 rounded-xl hover:bg-white-90 ${flexItemCenter}`;

  const handleToggleMenu = () => setIsOpenMenu(prev => !prev);
  const handleCloseMenu = () => setIsOpenMenu(false);
  const handleCloseSubMenu = () => setIsOpenSubMenu(false);

  const handleToggleSubMenu = () => setIsOpenSubMenu(prev => !prev);

  const handleCloseAllMenu = () => (handleCloseMenu(), handleCloseSubMenu());

  useEffect(() => {
    isOpenMenu
      ? (document.body.style.overflow = 'hidden')
      : (document.body.style.overflow = 'auto');
  }, [isOpenMenu]);

  return (
    <div>
      <BoxIcon
        icon={<BurgerMenuIcon />}
        onClick={handleToggleMenu}
        additionalClasses="w-12"
      />
      {isOpenMenu &&
        createPortal(
          <div className="w-full h-full z-10 absolute px-5 pb-4 pt-6 left-0 top-0 bg-white-100 shadow-xl">
            <div className="flex relative">
              <Brand brandName="MEGA.news" url="/" />
              <div className="absolute -right-2 -top-2">
                <BoxIcon
                  icon={<Close />}
                  onClick={handleCloseMenu}
                  additionalClasses="w-12"
                />
              </div>
            </div>
            <div
              className="flex justify-between items-center mt-4 h-12 pl-2 pr-3 rounded-xl hover:bg-white-90"
              onClick={handleToggleSubMenu}
            >
              <div className={`${flexItemCenter}`}>
                {isOpenSubMenu && (
                  <div className="w-1 h-[11px] bg-primary-100 rounded-2xl mr-2" />
                )}
                <Typography
                  tag="p"
                  textSize={Size.SM}
                  weight={FontWeight.Medium}
                  additionalClasses="text-dark-100 hover:text-primary-100"
                >
                  Categories
                </Typography>
              </div>
              {isOpenSubMenu ? <ArrowBottom /> : <ArrowRight />}
            </div>

            {isOpenSubMenu && (
              <ul className="list-none mt-1.5 pl-5">
                {listCategories.map(category => {
                  const { id, name } = category;
                  return (
                    <li
                      key={id}
                      className="mb-2 last:mb-0"
                      onClick={handleCloseAllMenu}
                    >
                      <Link href={`/category/${id}`}>
                        <Typography
                          tag="p"
                          textSize={Size.XS}
                          additionalClasses="text-dark-75 cursor-pointer hover:text-secondary-100"
                        >
                          { name }
                        </Typography>
                      </Link>
                    </li>
                  );
                })}
              </ul>
            )}

            <div>
              {listNavItems.map(navItem => (
                <div key={navItem.id} className={classNavItem}>
                  <NavItem name={navItem.name} url={navItem.path} />
                </div>
              ))}
            </div>
          </div>,
          document.body
        )}
    </div>
  );
};
