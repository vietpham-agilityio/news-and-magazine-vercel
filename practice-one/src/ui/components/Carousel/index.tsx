'use client';

// hooks
import { useState, useEffect, useCallback } from 'react';

// components
import Image from 'next/image';
import { BoxIcon, Typography } from '@/ui/components';

// types
import { FontWeight, ISlideItem, Size } from '@/types';

// icons
import { ArrowLeft, ArrowRight, Circle, Ellipse } from '@/ui/components/Icons';

interface IProps {
  listSlider: ISlideItem[];
  autoPlay?: boolean;
  autoPlayTime?: number;
}

export const Carousel = ({
  listSlider,
  autoPlay = true,
  autoPlayTime = 3000,
}: IProps) => {
  const [currentIndex, setCurrentIndex] = useState(0);

  const goToPrevious = useCallback(() => {
    const isFirstSlide = currentIndex === 0;
    const newIndex = isFirstSlide ? listSlider.length - 1 : currentIndex - 1;
    setCurrentIndex(newIndex);
  }, [currentIndex, listSlider.length]);

  const goToNext = useCallback(() => {
    const isLastSlide = currentIndex === listSlider.length - 1;
    const newIndex = isLastSlide ? 0 : currentIndex + 1;
    setCurrentIndex(newIndex);
  }, [currentIndex, listSlider.length]);

  useEffect(() => {
    if (autoPlay) {
      const interval = setInterval(goToNext, autoPlayTime);
      return () => clearInterval(interval);
    }
  }, [currentIndex, autoPlay, autoPlayTime, goToNext]);

  return (
    <div className="relative flex flex-col items-center w-full h-full mx-auto overflow-hidden border-2 bg-black-100 border-dark-10 rounded-lg">
      <BoxIcon
        width="10"
        height="10"
        additionalClasses="absolute left-4 z-10 top-1/2 hover:opacity-80"
        icon={<ArrowLeft />}
        onClick={goToPrevious}
      />
      <div
        className="w-full flex transition-transform duration-500 xl:w-126.5 2xl:w-186"
        style={{ transform: `translateX(-${currentIndex * 100}%)` }}
      >
        {listSlider.map((slider, index) => {
          const { image, title, content } = slider;

          return (
            <div
              key={index}
              className="w-full h-113 flex-shrink-0 flex flex-col items-center relative"
            >
              <Image
                fill
                objectFit='cover'
                key={index}
                src={image}
                alt={`Slide ${index}`}
              />
              <div className="absolute left-5 bottom-3.5 p-4 bg-white-75 rounded-xl opacity-80 w-[93%] lg:w-[96%] xl:w-[92%] 2xl:w-[95%]">
                <Typography
                  tag="h2"
                  textSize={Size.LG}
                  weight={FontWeight.Normal}
                  additionalClasses="text-black-100 h-10 text-ellipsis overflow-hidden"
                >
                  {title}
                </Typography>
                <Typography
                  tag="h3"
                  textSize={Size.XSM}
                  weight={FontWeight.Light}
                  additionalClasses="text-dark-100 w-100 lg:w-full xl:w-75 2xl:w-[543px] h-10 leading-5 text-ellipsis overflow-hidden"
                >
                  {content}
                </Typography>
              </div>
            </div>
          );
        })}
      </div>
      <BoxIcon
        width="10"
        height="10"
        additionalClasses="absolute right-4 top-1/2 hover:opacity-80"
        icon={<ArrowRight />}
        onClick={goToNext}
      />

      <div className="flex absolute bottom-4 right-6">
        {listSlider.map((_, index) => (
          <button
            key={index}
            aria-label="slider-button"
            className="h-3 z-10 rounded-full w-6"
            onClick={() => setCurrentIndex(index)}
          >
            {index === currentIndex ? <Ellipse /> : <Circle />}
          </button>
        ))}
      </div>
    </div>
  );
};
