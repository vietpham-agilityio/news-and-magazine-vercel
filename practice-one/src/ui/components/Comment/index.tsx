// components
import { FontWeight, Size } from "@/types";
import { Typography } from "@/ui/components/common";

interface IProps {
  userName: string;
  content: string;
}

export const Comment = ({ userName, content }: IProps) => (
  <div className="bg-white-90 p-4 rounded-xl">
    <Typography tag="h2" textSize={Size.SM} additionalClasses="text-dark-100">
      { userName }
    </Typography>
    <Typography
      tag="h3"
      weight={FontWeight.Light}
      additionalClasses="text-p text-dark-75"
    >
      { content }
    </Typography>
  </div>
);
