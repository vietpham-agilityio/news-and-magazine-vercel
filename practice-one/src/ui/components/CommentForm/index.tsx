// components
import { Button, Input, TextArea } from '@/ui/components';

// icons
import { ChatBoxSmall } from '@/ui/components/Icons';

export const CommentForm = () => (
  <div>
    <div className="grid grid-cols-12 grid-rows-3 lg:gap-6">
      <div className="row-span-3 col-span-12 lg:col-span-6">
        <div className="mb-5 row-span-1">
          <Input label="Name" name="username" type="text" />
        </div>
        <div className="mb-5 row-span-1">
          <Input label="Email" name="email" type="email" />
        </div>
        <div className="mb-5 row-span-1">
          <Input label="Website" name="website" type="text" />
        </div>
      </div>
      <div className="row-span-3 col-span-12 lg:col-span-6">
        <TextArea label="Comment" placeholder="What do you think?" />
      </div>
    </div>
    <div className="flex w-full justify-end mt-7 mb-30 sm:mt-10 sm:mb-12 lg:mt-0">
      <Button
        isButtonLarge
        label="Send Comment"
        leftIcon={<ChatBoxSmall />}
        bgColor="bg-gradient-to-r from-sunSet-0 to-sunSet-100"
        textColor="text-white-100"
      />
    </div>
  </div>
);
