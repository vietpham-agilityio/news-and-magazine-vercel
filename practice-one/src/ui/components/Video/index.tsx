import Image, { StaticImageData } from 'next/image';
import { Typography } from '@/ui/components';

// types
import { FontWeight, Size } from '@/types';
import { Triangle } from '../Icons';

interface IProps {
  title: string;
  content: string;
  videoSrc: string | StaticImageData;
}

export const Video = ({ title, content, videoSrc }: IProps) => (
  <div className="h-full w-full relative rounded-xl flex justify-center bg-transparent p-2.5 hover:cursor-pointer">
    <Image
      className="rounded-xl absolute"
      src={videoSrc} // will replace image component with video
      alt="panel of video"
      fill
    />
    {/* card body */}
    <div className="hidden sm:block absolute left-2 xl:left-2.5 bottom-4 p-4 w-[97%] bg-white-75 rounded-xl opacity-90">
      <Typography
        tag="h2"
        textSize={Size.MD}
        weight={FontWeight.Medium}
        additionalClasses="text-black-100"
      >
        {title}
      </Typography>
      <Typography
        tag="h3"
        textSize={Size.XSM}
        weight={FontWeight.Light}
        additionalClasses="text-dark-100 h-10 w-full pr-10 leading-5 text-ellipsis overflow-hidden "
      >
        {content}
      </Typography>
    </div>
    <div className="absolute flex items-center justify-center w-30 h-30 mt-32 sm:mt-24 bg-white-90 opacity-85 rounded-full hover:cursor-pointer">
      <div className="ml-1">
        <Triangle />
      </div>
    </div>
  </div>
);
