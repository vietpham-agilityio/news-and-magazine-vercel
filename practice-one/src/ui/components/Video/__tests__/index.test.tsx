import { render, cleanup } from '@testing-library/react';

// component
import { Video } from '@/ui/components';

let renderVideo: any;
const title: string ="How Music Affects Your Brain (Plus 11 Artists To Listen To At Work)"
const content: string ="Youâ€™ve Read All Your Free Member-Only Stories, Become A Member To Get Unlimited Access. Your Membership Fee Supports The Voices You Want To Hear More From."
const videoSrc: string ="https://images.unsplash.com/photo-1717496002081-38a90d3cba26?w=500&auto=format&fit=crop&q=60&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8"


describe('Video component', () => {
  beforeEach(() => {
    renderVideo = render(<Video videoSrc={videoSrc} title={title} content={content} />);
  });

  afterEach(() => {
    renderVideo.unmount();
    cleanup();
  });

  it('should render Video macth snapshot', () => {
    expect(renderVideo.asFragment()).toMatchSnapshot();
  });

  it('Video should render value passed title props', () => {
    const { getByText } = renderVideo

    const titleElement = getByText(title);

    expect(titleElement).toBeInTheDocument();
  });

 
  it('Video should render value passed content props', () => {
    const { getByText } = renderVideo

    const contentElement = getByText(content);

    expect(contentElement).toBeInTheDocument();
  });
});
