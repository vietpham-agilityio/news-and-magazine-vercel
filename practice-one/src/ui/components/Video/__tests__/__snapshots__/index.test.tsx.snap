// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video component should render Video macth snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full w-full relative rounded-xl flex justify-center bg-transparent p-2.5 hover:cursor-pointer"
  >
    <img
      alt="panel of video"
      class="rounded-xl absolute"
      data-nimg="fill"
      decoding="async"
      loading="lazy"
      sizes="100vw"
      src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=3840&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1717496002081-38a90d3cba26%3Fw%3D500%26auto%3Dformat%26fit%3Dcrop%26q%3D60%26ixlib%3Drb-4.0.3%26ixid%3DM3wxMjA3fDB8MHxlZGl0b3JpYWwtZmVlZHw2fHx8ZW58MHx8fHx8&w=3840&q=75 3840w"
      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
    />
    <div
      class="hidden sm:block absolute left-2 xl:left-2.5 bottom-4 p-4 w-[97%] bg-white-75 rounded-xl opacity-90"
    >
      <h2
        aria-level="2"
        class="text-md font-medium text-black-100"
        role="heading"
      >
        How Music Affects Your Brain (Plus 11 Artists To Listen To At Work)
      </h2>
      <h3
        aria-level="3"
        class="text-xsm font-light text-dark-100 h-10 w-full pr-10 leading-5 text-ellipsis overflow-hidden "
        role="heading"
      >
        Youâ€™ve Read All Your Free Member-Only Stories, Become A Member To Get Unlimited Access. Your Membership Fee Supports The Voices You Want To Hear More From.
      </h3>
    </div>
    <div
      class="absolute flex items-center justify-center w-30 h-30 mt-32 sm:mt-24 bg-white-90 opacity-85 rounded-full hover:cursor-pointer"
    >
      <div
        class="ml-1"
      >
        <svg
          data-testid="triangle-icon"
          fill="none"
          height="44"
          width="36"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M33.844 18.156C35.156 19 36 20.5 36 22c0 1.594-.844 3.094-2.156 3.844l-27 16.5c-1.407.843-3.188.937-4.594.093C.844 41.688 0 40.188 0 38.5v-33c0-1.594.844-3.094 2.25-3.844a4.285 4.285 0 0 1 4.594.094l27 16.406Z"
            fill="#F81539"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
